### PR TITLE
Refactored Autocomplete

### PR DESCRIPTION
**Pull request by untra**
* changes to snippets to predict and suggest their prefixes with their bodies
* changed the tsconfig javascript library to `es7` from `es6`
* removed some extra code from  `elixirSenseDefinitionProvider`
* removed an extraneous print statement from the `elixirSenseSignatureProvider`
* crazy sexy refactor of `elixirSenseAutocompleteProvider`
  * simplified the code considerably. Removed a number of existing rules. If there's a particular feature you're looking for, be sure to test for it!
  * callbacks get a new symbol, instantly apply the new `@impl` attribute, and autocomplete with an empty snippet :+1:
  * macros get a new symbol from functions
  * labels now display with their arity
  * kernel functions and macros starting with a symbol are operators, and get a new symbol

### CODE REVIEW

**Comment by untra** [comment]
Okay, the dust has settled. This is a continuation of #83, which I mangled pretty hard. Lets see if this PR takes. @mat-mcloughlin do you want to take a look?

File: src/elixirSenseClient.ts [hunk 1]
```diff
@@ -50,6 +52,13 @@ export class ElixirSenseClient {
     handleConnect() {
         this.resetBuffer(0, 4);
         console.log(`[vscode-elixir] ElixirSense client connected on ${this.host}:${this.port}`);
+        this.send('version', {})
```
Comment (untra) [inline src/elixirSenseClient.ts]: awaiting a change in https://github.com/msaraiva/elixir_sense/pull/8 , to use this endpoint

**Comment by mat-mcloughlin** [comment]
@untra yeah I'll take a look today. Whats the advantage of elixirsense over alchemist?

**Comment by untra** [comment]
Alchemist-Server hasn't gotten as much love as elixirsense has, and the repo contents were copied over ~6 months ago, rather than git-submoduled like elixirsense. I also prefer elixirsense's api and integration, and have significantly cleaned up the implementation around elixirsense components to use [promise chains](https://github.com/fr1zle/vscode-elixir/pull/90/files#diff-7591485bae15ac02d0a3b1808e688fa3R34) rather than complex tangled callbacks.

Which language server gets used is toggleable. I made updates to the elixirsense server because it is default, and it gets the most use 👍

**Comment by mat-mcloughlin** [comment]
yeah always meant to submodule the alchemist stuff off but never got around to it. 

I'll have to have a look in more detail. Thanks

File: src/elixirSenseClient.ts (hunk 1 above)
Comment (fr1zle) [reply src/elixirSenseClient.ts]: So, its probably best to wait for that and not pull this PR for now?

Comment (untra) [reply src/elixirSenseClient.ts]: It is in! I'm going to whip up a quick pr to update the submodule to elixirsense v1.0.0

**Comment by untra** [comment]
Alright, this PR has the updated version elixirSense in it! All the endpoints should now work as intended. give it a play through, and merge it in when you're ready! 👍

**Comment by fr1zle** [comment]
@untra any reason why you closed this?

**Comment by untra** [comment]
My apologies. Reopened.

**Comment by fr1zle** [comment]
Going to merge this for now, thank you.

I have seen some crashes, but haven't been able to reproduce them after restarting, yet.

Will test a bit more and then prepare a release.

**Comment by fr1zle** [comment]
@untra i see a lot of exceptions when navigating in testcase, probably because we start the elixir_sense server in the "dev" environment. So i don't think this was caused by this PR.

Maybe we can find a way to get all editor features working in testcases aswell... maybe by changing the default to "test"?

**Comment by untra** [comment]
I'll tinker around with this further. I did notice a few exceptions, but they seem to be related to elixir sense rather than the plugin. There are still some other changes / fixes with elixirsense, and upcoming versions of elixir/otp promise full support for accesible erlang documentation (which will mean further changes to elixir sense and this plugin).

Thanks for getting this merged in! 👍

**Pull request merged**